draft-ietf-tls-esni.md

```diff
@@ -787,34 +785,35 @@ of the "ech_required" alert to detect this case.
 
 ### HelloRetryRequest {#server-hrr}
 
-In case a HelloRetryRequest (HRR) is sent, the client-facing server MUST
-consistently accept or decline ECH between the two ClientHellos, using the same
-ECHConfig, and abort the handshake if this is not possible. This is achieved as
-follows. Let CH1 and CH2 denote, respectively, the first and second ClientHello
-transmitted on the wire by the client:
+After sending a HelloRetryRequest, the client-facing server does not repeat the
+steps in {{client-facing-server}} with the second ClientHelloOuter. Instead it
+continues the selections from the first ClientHelloOuter as follows:
 
-1. If CH1 contains the "encrypted_client_hello" extension but CH2 does not, or
-   if CH2 contains the "encrypted_client_hello" extension but CH1 does not, then
-   the server MUST abort the handshake with an "illegal_parameter" alert.
-1. If the "encrypted_client_hello" extension is sent in CH2, the server follows
-   the procedure in {{client-facing-server}} to decrypt the extension, but it
-   uses the previously-selected ECHConfig as the set of candidate ECHConfigs.
-   If decryption fails, the server aborts the connection with a "decrypt_error"
-   alert rather than continuing the handshake with the second ClientHelloOuter.
-
-When decrypting the second ClientECH.payload, the client-facing server
-performs a corresponding process to {{client-hrr}}:
+If the client-facing server accepted ECH, it checks the second ClientHelloOuter
+also contains the "encrypted_client_hello" extension. If not, it MUST abort the
+handshake with a "missing_extension" alert. Otherwise, it decrypts the new
+ClientECH.payload as a second message with the previous HPKE context:
```

Comment:
As long as "cipher_suite", "config_id", "enc" are retransmitted in CH2, should the server verify that they don't change?

Reply:
Checking CH1 and CH2 consistency is a huge mess to do. We completely got the design of HRR wrong in RFC8446, and it's too late to do anything about that now. Rather than expend a whole lot of energy on everyone's part, from spec text to implementations to constantly having this discussion, I'd rather we follow RFC8446's example here and leave it alone.

Reply:
That is, RFC8446 is already quite restrictive about CH1/CH2 consistency, yet servers are not required to (and in practice do not) check consistency for much, if any, of it.
https://tools.ietf.org/html/rfc8446#section-4.1.2

Indeed if you think about what a server would need to do here, it's clear we just got the design wrong. A server that is capable of checking some individual field most likely stored the value of the field, in which case there was no reason for the client to re-send it in the first place! (This grew organically from an earlier design where the server actually reconstructed CH1 from CH2, but that had all kinds of attacks and was generally a mess. When we dropped that, we probably should have stopped to reevaluate it. Oh well.)

Reply:
This seems like it's orthogonal, right?

Reply:
Well, the question is whether we mandate the server check the redundant fields. I think my default is to prefer matching rfc8446 in not saying much. (With a side of grumpiness about how weird HRR is. :-) ) We probably need to do a better job describing HRR in rfc8446bis. I figure the less other specs say about it, the more room we have to figure out what we actually want to require.

Reply:
(Supposing we go with https://github.com/tlswg/draft-ietf-tls-esni/pull/352#discussion_r518868173, the redundant fields are gone and this is moot.)